plugin/federated/federated_plugin.h

```diff
@@ -0,0 +1,361 @@
+ *
+ *  Symbol name `FederatedPluginBuildEncryptedHistHori`.
+ *
+ * @param in_hist  The input local histogram.
+ * @param len      The size of the input local histogram.
+ * @param out_hist Encrypted histogram.
+ * @param out_len  The size of the encrypted histogram.
+ *
+ * @return 0 if succees.
+ */
+using BuildHistHoriFn = int(FederatedPluginHandle handle, double const *in_hist, size_t len,
+                            uint8_t **out_hist, size_t *out_len);
+/**
+ * @brief Reduce the histogram after the allgather call.
+ *
+ *  Symbol name: `FederatedPluginSyncEnrcyptedHistHori`.
+ *
+ * @param in_hist  Input histogram from the allgather call.
+ * @param len      The length of the input histogram.
+ * @param out_hist Output histogram.
+ * @param out_len  The size of the output histogram.
+ *
+ * @return 0 if succees.
+ */
+using SyncHistHoriFn = int(FederatedPluginHandle handle, uint8_t const *in_hist, size_t len,
+                           double **out_hist, size_t *out_len);
+/**@}*/
+}  // namespace federated
+
+// Base class for federated learning plugin.
+class FederatedPluginBase {
+  std::vector<std::uint8_t> grad_;
+  std::vector<std::uint8_t> hist_enc_;
+  std::vector<double> hist_plain_;
+
+ public:
+  [[nodiscard]] virtual common::Span<std::uint8_t> EncryptGradient(
+      common::Span<float const> data) = 0;
+  virtual void SyncEncryptedGradient(common::Span<std::uint8_t const>) = 0;
+
+  // Vertical histogram
+  virtual void Reset(common::Span<std::uint32_t const>, common::Span<std::int32_t const>) {}
+
+  [[nodiscard]] virtual common::Span<std::uint8_t> BuildEncryptedHistVert(
+      common::Span<std::uint64_t const *> rowptrs, common::Span<std::size_t const> sizes,
+      common::Span<bst_node_t const> nids) = 0;
+
+  [[nodiscard]] virtual common::Span<double> SyncEncryptedHistVert(
+      common::Span<std::uint8_t> hist) = 0;
+
+  // Horizontal histogram
+  [[nodiscard]] virtual common::Span<std::uint8_t> BuildEncryptedHistHori(
+      common::Span<double const> hist) = 0;
+  [[nodiscard]] virtual common::Span<double> SyncEncryptedHistHori(
+      common::Span<std::uint8_t const> hist) = 0;
+};
+
+// Only used for testing, this class is an no-op implementation.
+class FederatedPluginMock : public FederatedPluginBase {
+  Context ctx_;
+  std::vector<std::uint8_t> grad_;
+
+  std::vector<std::uint8_t> hist_enc_;  // represents the encrypted histogram
+  std::vector<double> hist_plain_;      // represents the plain text histogram
+
+  std::vector<std::uint32_t> cuts_;  // HistogramCuts::Ptrs()
+  GHistIndexMatrix gmat_;
+
+ public:
+  ~FederatedPluginMock() = default;
+
+  [[nodiscard]] common::Span<std::uint8_t> EncryptGradient(
+      common::Span<float const> data) override {
+    grad_.resize(data.size_bytes());
+    auto casted =
+        common::Span{reinterpret_cast<std::uint8_t const *>(data.data()), data.size_bytes()};
+    std::copy_n(casted.data(), casted.size(), grad_.data());
```

Comment:
Do we still plan to use the mock plugin for unencrypted federated learning? I wonder if we can somehow avoid these types of copying.

Reply:
Not at the moment, it's only used for running tests. Any suggestions? I don't what this mock myself, but haven't found a way to run test without one.

Reply:
Probably we can write the histogram build in a way that can reuse the existing histogram implementation. Will look into it tomorrow.

Reply:
Hmm, then the tests wouldn't be reliable. Also, the plugin itself cannot communicate, so we can't hand the entire histogram build to the plugin. 
Any suggestion for design would be great.

Reply:
For testing it's probably fine.